module-iam.tf

```diff
@@ -2,7 +2,7 @@
 # Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracle.com/licenses/upl
 
 data "oci_identity_availability_domains" "all" {
-  compartment_id = local.compartment_id
+  compartment_id = local.tenancy_id != "unknown" ? local.tenancy_id : local.compartment_id
```

Comment:
This one is more a limitation of the OCI API and the Terraform provider. If the `compartment_id` is not known during planning, then Terraform does not know how many `data.oci_identity_availability_domains.all.availability_domains` will exist. When eventually used as a `for_each` for `module.workers.data.oci_identity_fault_domains.all`, it will fail.

Best option is to try to use the static Tenancy OCID if possible and only fallback to the Compartment OCID if that is unknown

Reply:
The one issue here is for some scenarios where auth != resource tenancy, and the returned AD prefixes are incorrect when the same provider value is used.

Another workaround I can think of might be to define a map local with 3 static entries each conditionally set if available from the data source, so the count is known?

Reply:
@devoncrouse Is the scenario you described that if the provider auth does not have access to the whole tenancy and is restricted to a subcompartment, then listing the availability domains against the tenancy OCID fails? Or do you mean there is a scenario where the subcompartment's ADs differ from the parent tenancy's ADs?

Regarding the local map, if I understood correctly, I still don't think it fully resolves the dependency. If the value of the local map depends on a data source, and that map gets used in a `for_each` or `count`, Terraform will still be relying on the data source to know the count. Only similar approach I can see is just hard-coding `FD-1/2/3` and not bothering with `data.oci_identity_fault_domains` at all. Let me know if I'm missing something

Reply:
@devoncrouse to cover the use case you described we should be more specific in the example we provide for the provider authentication.

`tenancy_id`/`tenancy_ocid` should be used to target the resource tenancy.
`auth_tenancy_id` should be used to override the above in the provider configuration.

Most of the users of this module consider `tenancy_id` as the resource tenancy and for me, it makes sense to use it for ADs name.